### ISSUE TITLE
Arc-ed assets

### ISSUE THREAD

**Issue by andriyDev** [bevyengine/bevy#15723]
## What problem does this solve or what need does it fill?

Today, there is no way to access an asset from a thread/async block. The best you can do is clone the asset entirely, or make your asset a wrapper around an Arc with all the data. This is especially painful for https://github.com/bevyengine/bevy/issues/11216, which needs to be able to send assets (and their subassets) away to be serialized.

## What solution would you like?

Store assets as `Arc<A>`! Now the asset can be passed to threads and read from there.

### Design proposal

1. In `assets::Entry<A>`, instead of storing an `Option<A>`, store an `Option<Arc<A>>`.
2. Make `Assets<A>::get` return an `Option<Arc<A>>`
3. Remove `Assets<A>::get_mut`.
4. Make `Assets<A>::insert` and `Assets<A>::add` take `impl Into<Arc<A>>` instead of `impl Into<A>`.

    This is a loss in functionality (you can't pass in something that converts to an `A`), but I don't think it's a big loss.

5. Make `AssetEvent::Modified` occur whenever an asset is replaced.

### Optional additional steps

1. Make the render world take `Arc`s instead of cloning.

    This improves the render asset situation *for assets that are both in `MAIN_WORLD` and `RENDER_WORLD`*. This does not matter for assets that must be only in the `RENDER_WORLD` since moving the assets has to happen anyway. There's a slight win in that moving the assets is now just a pointer copy, but there's also a slight loss in that creating `RenderAssets` may need to do an extra clone if they want to move parts of a vector or something to the render asset.

2. Implement `Assets<A>::get_cloned_mut`.

    Since assets are no longer mutable (due to the `Arc`), we can provide a utility function for assets that implement `Clone` that just calls `Arc::make_mut`. This allows users to seem like they are mutating an asset even though they're really cloning on write (when necessary; turns out `make_mut` will only clone if there are other `Arc`s, quite neat!).

3. Consider deleting `DenseAssetStorage`

    Today, `DenseAssetStorage` is used because we store assets densely to improve cache hits and stuff. With `Arc`s this probably doesn't matter much (since accessing assets will require a pointer hop anyway). So we could replace `DenseAssetStorage` with a `HashMap`. In fact we could merge it with the Uuid `HashMap`, making `Assets` basically a wrapper around `HashMap<AssetId, Arc<A>>`. In this model, `AssetIndex` will probably just become a u64 that counts up.

## What alternative(s) have you considered?

Asset locking - #15346

### Advantages to `Arc`ing

- The "next steps" section shows many simplifications we can implement. Less code is preferable!
- Assets in an async context are the "front door" - just pass around the `Arc`!

### Disadvantages to `Arc`ing

- It becomes very easy for users to just clone the `Arc` and then never "poll" the `Assets` resource again. This makes it a hazard for hot-reloading or any case where you want mutable assets. Asset locking is a scary enough name that users are more likely to avoid it by default and only reach for it when necessary (and may also think about how to unlock the asset).

**Comment by djeedai** [comment]
@alice-i-cecile This removes asset mutability entirely (see point 3.). It doesn't look like there's any discussion about it nor even any mention of this drastic change of behavior. This will break not only asset reloading but also _e.g._ tweening/animation. I think this should at the very least have a Controversial label. I see commits being made against this, and I doubt this should merge or even work being done at all before we clarified that design. Based on the explanation in this issue, and the understanding I'm making of it as a result, I'm completely against this change.

**Comment by andriyDev** [comment]
@djeedai the associated PR may shed more light on this. Namely, it does implement `get_cloned_mut` and also adds `get_inplace_mut`. Both of these only change behavior from existing `get_mut` if something is holding the arc. Otherwise, they just give you mutable access as normal.

Is this not sufficient? This are absolutely trade-offs here, mostly in terms of complexity, but I don't agree that we're losing much in terms of functionality.

**Comment by djeedai** [comment]
What happens when some game data holds an `Arc`, and you use asset hot-reloading? As far as I understand, your change creates a new copy and the reloaded asset data gets applied to that copy that no one uses. This effectively breaks hot-reloading.

More generally, `Arc` and `AssetId` serve the same role of being a handle to shared data. In fact `Handle` even has a strong variant and a weak one. You can build an asset system with one or the other; now we're duplicating them and proposing to use both at the same time.

**Comment by andriyDev** [comment]
Yes, there is an assumption being made here that the Arcs are being held for *short* periods of time. I would argue that doesn't break hot reloading. We still expect users to "return to the ECS" regularly to get the asset data/asset arc. With a big enough warning comment on fetching the arc, I don't see this as a significant problem.

I don't really agree that these serve the same role. I view the arc more as a reference than an identifier. The same way that when you `Assets::get` an asset, you get back a reference, `Assets::get_arc` gets its own "kind" of reference.
